dataactcore/scripts/loadHistoricalFabs.py

```diff
@@ -177,6 +187,141 @@ def format_proper_casing(row, header):
     return row[header]
 
 
+def derive_legal_entity_city_code(row, sess):
+    if row['recipient_city_code'] and not is_nan(row['recipient_city_code']):
+        return row['recipient_city_name']
+    elif row['recipient_city_name'] and row['recipient_state_code'] and not is_nan(row['recipient_city_name']) \
+            and not is_nan(row['recipient_state_code']):
+        city_data = sess.query(CityCode). \
+            filter(func.lower(CityCode.feature_name) == func.lower(row['recipient_city_name'].strip()),
+                   func.lower(CityCode.state_code) == func.lower(
+                       row['recipient_state_code'].strip())).one_or_none()
+        if city_data:
+            return city_data.city_code
+    return None
+
+
+def derive_awarding_agency_code(row, sess):
+    if not row['agency_code']:
+        return None
+    awarding_sub_tier = sess.query(SubTierAgency). \
+        filter_by(sub_tier_agency_code=row['agency_code']).one()
+    use_frec = awarding_sub_tier.is_frec
+    awarding_agency = awarding_sub_tier.frec if use_frec else awarding_sub_tier.cgac
+    return awarding_agency.frec_code if use_frec else awarding_agency.cgac_code
+
+
+def derive_awarding_agency_name(row, sess):
+    if not row['agency_code']:
+        return None
+    awarding_sub_tier = sess.query(SubTierAgency). \
+        filter_by(sub_tier_agency_code=row['agency_code']).one()
+    use_frec = awarding_sub_tier.is_frec
+    awarding_agency = awarding_sub_tier.frec if use_frec else awarding_sub_tier.cgac
+    return awarding_agency.agency_name
+
+
+def derive_awarding_sub_tier_agency_n(row, sess):
+    if not row['agency_code']:
+        return None
+    awarding_sub_tier = sess.query(SubTierAgency). \
+        filter_by(sub_tier_agency_code=row['agency_code']).one()
+    return awarding_sub_tier.sub_tier_agency_name
+
+
+def derive_place_of_perform_county_na(row, sess):
+    if not row['principal_place_zip']:
+        ppop_code = row['principal_place_code'].upper()
+        ppop_state = None
+        if ppop_code[:2] == '00':
+            return None
+        elif re.match('^[0-9]{2}', ppop_code[:2]):
```

Comment:
If we're only grabbing the first 2 characters, do we need the "start" character on this regex?

Reply:
no